libgadget/domain.c

```diff
@@ -377,22 +418,22 @@ domain_balance(DomainDecomp * ddecomp)
     walltime_measure("/Domain/Decompose/Sumcost");
 
     /* first try work balance */
-    domain_assign_balanced(ddecomp, TopLeafWork);
+    domain_assign_balanced(ddecomp, TopLeafWork, NTask, NTask);
```

Comment:
This looks strange. It looks like you wanted NSubDomains or NSegment here.

Reply:
NSubDomains is used only at an earlier stage of the process, to build more TopLeaves than we need Segments. The assign_balanced code then multiple TopLeaves to Segments so that each Segment is of roughly equal weight, and then assigns these Segments to Tasks so that the Tasks are also of roughly equal weight.

The thing is, in the current algorithm the Segments are superfluous: because each Segment is of equal weight anyway there is no point making them separate from Tasks. This is specifying that a Segment and a Task are the same. I could simplify by removing the code for multiple Segments per Task but I don't want to yet because the (finely-tuned) algorithm may still turn out to be useful.

Reply:
The new code has this clarified.